async-book/src/execution/executor.md

```diff
@@ -0,0 +1,160 @@
+# Applied: Build an Executor
+
+For this one, we're going to have to include the `futures` crate in order to
+get the `FutureObj` type, which is a dynamically-dispatched `Future`, similar
+to `Box<dyn Future<Output = T>>`. `Cargo.toml` should look something like this:
+
+```toml
+[package]
+name = "xyz"
+version = "0.1.0"
+authors = ["XYZ Author"]
+edition = "2018"
+
+[dependencies]
+futures-preview = "0.3.0-alpha.9"
+```
+
+Next, we need the following imports at the top of `src/main.rs`:
+
+```rust
+#![feature(arbitrary_self_types, async_await, await_macro, futures_api, pin)]
+
+use {
+    futures::future::FutureObj,
+    std::{
+        future::Future,
+        pin::Pin,
+        sync::{Arc, Mutex},
+        sync::mpsc::{sync_channel, SyncSender, Receiver},
+        task::{
+            local_waker_from_nonlocal,
+            Poll, Wake,
+        },
+    },
+};
+```
+
+Our executor will work by sending tasks to run over a channel. The executor
+will pull events off of the channel and run them. When a task is ready to
+do more work (is awoken), it can schedule itself to be polled again by
+putting itself back onto the channel.
+
+In this design, the executor itself just needs the receiving end of the task
+channel. The user will get a sending end so that they can spawn new futures.
+Tasks themselves are just futures that can reschedule themselves, so we'll
+store them as a future paired with a sender that the task can use to requeue
+itself.
+
+```rust
+/// Task executor that receives tasks off of a channel and runs them.
+struct Executor {
+    task_receiver: Receiver<Arc<Task>>,
+}
+
+/// `Spawner` spawns new futures onto the task channel.
+#[derive(Clone)]
+struct Spawner {
+    task_sender: SyncSender<Arc<Task>>,
+}
+
+/// A future that can reschedule itself to be polled using a channel.
+struct Task {
+    // In-progress future that should be pushed to completion
+    future: Mutex<Option<FutureObj<'static, ()>>>,
```

Comment:
You never explain why the `Mutex` is needed here.

Reply:
I'm guessing "to make rustc happy" isn't the answer you're looking for

Reply:
Added some text here-- lemme know what you think :)